### PR TITLE
feat(actions): expand all under folder

### LINKED ISSUE #1289: expand_all to expand all folders to make live_filter more awesome
**Is your feature request related to a problem? Please describe.**

The new feature "live filter" is pretty awesome but it is not very usable due to two reasons:

1. It only searches expanded folders (I actually think this is desirable). 
2. There is no shortcut for expanding all folders

**Describe the solution you'd like**

I think it'd make sense to add a shortcut to expand all folders. There is already one to colapse_all but not to do the opposite.

**Describe alternatives you've considered**

N/A

**Additional context**

N/A

### PR DESCRIPTION
**Pull request by kyazdani42**
fixes #1289 #380 
@alex-courtis i'd like a review on the doc and the binding to make sure its okay, because this feature will freeze neovim if used in the wrong place.

### CODE REVIEW

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Looks good, works.

Froze for 15 mins when expanding `~/src` but that is not surprising.

Needs a safety valve e.g.
"Expand halted after opening 500 folders"
That could be hardcoded or an option.

**Comment by kyazdani42** [comment]
i think 100/200 folders would already be good enough, i'll try something out

File: README.md [hunk 1]
```diff
@@ -190,6 +190,9 @@ require'nvim-tree'.setup { -- BEGIN_DEFAULT_OPTS
       global = false,
       restrict_above_cwd = false,
     },
+    expand_all = {
+      max_folder_discovery = 300,
```
Comment (kyazdani42) [inline README.md:194]: not too sure about the option name here.

File: lua/nvim-tree/actions/expand-all.lua [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    end
+  end
+
+  return iterate
+end
+
+function M.fn(base_node)
+  local node = base_node.nodes and base_node or core.get_explorer()
+  if gen_iterator()(node) then
+    utils.warn("expansion iteration was halted after " .. M.MAX_FOLDER_DISCOVERY .. " discovered folders")
```
Comment (kyazdani42) [inline lua/nvim-tree/actions/expand-all.lua:47]: not sure if this is needed.

File: README.md (hunk 1 above)
Comment (alex-courtis) [reply README.md:194]: `max_folder_expand` ?

File: lua/nvim-tree/actions/expand-all.lua (hunk 2 above)
Comment (alex-courtis) [reply lua/nvim-tree/actions/expand-all.lua:47]: I think it is good to tell the user what is happening.

**Comment by alex-courtis** [comment]
> i think 100/200 folders would already be good enough, i'll try something out

300 is good; it stops after a few seconds.

File: README.md [hunk 3]
```diff
@@ -261,6 +264,7 @@ require'nvim-tree'.setup { -- BEGIN_DEFAULT_OPTS
 - Double left click acts like `<CR>`
 - Double right click acts like `<C-]>`
 - `W` will collapse the whole tree
+- `E` will expand the whole tree. Be aware this might hang neovim for a while if running on a big folder (such as home dir or root dir).
```
Comment (alex-courtis) [inline README.md:267]: Reference `actions.expand_all.max_folder_discovery` here.

**Review by alex-courtis — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (kyazdani42) [reply README.md:194]: we'll see if this makes sense for user as it is.

**Pull request merged**